src/libcore/iter/mod.rs

```diff
@@ -499,6 +499,18 @@ unsafe impl<'a, I, T: 'a> TrustedRandomAccess for Cloned<I>
     fn may_have_side_effect() -> bool { true }
 }
 
+#[doc(hidden)]
+unsafe impl<'a, I, T: 'a> TrustedRandomAccess for Cloned<I>
```

Comment:
Why is this specialization required? It seems totally equivalent to the impl above.

Reply:
It returns `false` for `may_have_side_effect`.

Reply:
Maybe may_have_side_effect could be removed and it just assumes everything has side effects. Since some .map().zip() invocations optimize ok it could be suspected that the compiler can remove dead code itself. Unfortunately a closer examination of benchmarks is needed.

Reply:
Ah right.

Reply:
This does assume that `Clone` is simply `*self`, but it could do weird side-effectful things. I think it's probably reasonable to state that we're allowed to assume that `Clone` behaves sanely for `Copy` types, but I don't think that's explicitly documented currently.

Reply:
It appears to be in fact documented: https://doc.rust-lang.org/std/clone/trait.Clone.html#how-can-i-implement-clone

Reply:
Ah great, I missed that.